Add dotnet/roslyn-analyzers repo

omajid opened the pull request:
This repo doesn't build out of the box on Linux, so I have added various build scripts and hacks to get that working.

omajid commented on patches/roslyn-analyzers/0001-WIP-for-source-build.patch:
```diff
@@ -0,0 +1,1087 @@
+ create mode 100644 src/Text.Analyzers/Setup/AnalyzerReleases.Unshipped.md
+
+diff --git a/FxCopAnalyzers.sln b/FxCopAnalyzers.sln
+deleted file mode 100644
```
Is there some other way to selectively disable a `.sln` file?

dagood replied:
Maybe we can specify one in the build command... but I think we can defer this improvement.

omajid commented on patches/roslyn-analyzers/0001-WIP-for-source-build.patch:
```diff
@@ -0,0 +1,1087 @@
+ EndProject
+ Project("{778DAE3C-4631-46EA-AA77-85C1314464D9}") = "Roslyn.Diagnostics.VisualBasic.Analyzers", "src\Roslyn.Diagnostics.Analyzers\VisualBasic\Roslyn.Diagnostics.VisualBasic.Analyzers.vbproj", "{70BBA457-2CC1-4929-8FEE-359EBB7C398A}"
+ EndProject
+-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Roslyn.Diagnostics.Analyzers.UnitTests", "src\Roslyn.Diagnostics.Analyzers\UnitTests\Roslyn.Diagnostics.Analyzers.UnitTests.csproj", "{D90E7402-70E7-4A95-A292-47EA163D6309}"
```
I have removed all Unit Test projects. Not sure if we should remove fewer or more projects.

omajid commented on patches/roslyn-analyzers/0001-WIP-for-source-build.patch:
```diff
@@ -0,0 +1,1087 @@
++done
++
++scriptroot="$( cd -P "$( dirname "$source" )" && pwd )"
++"$scriptroot/eng/common/build.sh" --restore --build $@
```
This script is basically a port of [`Build.cmd`](https://github.com/dotnet/roslyn-analyzers/blob/master/Build.cmd) which really just calls `eng/common/build.ps1 --restore --build`.

omajid commented on patches/roslyn-analyzers/0001-WIP-for-source-build.patch:
```diff
@@ -0,0 +1,1087 @@
+       <_CscToolPath Condition="!HasTrailingSlash('$(_CscToolPath)')">$(_CscToolPath)\</_CscToolPath>
+     </PropertyGroup>
+ 
+-    <Exec Command='"$(_CscToolPath)csi.exe" "$(RepoRoot)eng\GenerateAnalyzerNuspec.csx" "$(NuspecFile)" "$(AssetsDir)\" "$(MSBuildProjectDirectory)" "$(Configuration)" "$(TargetFrameworksForPackage)" "@(_NuspecMetadata)" "@(AnalyzerNupkgFile)" "@(AnalyzerNupkgFolder)" "@(AnalyzerNupkgAssembly)" "@(AnalyzerNupkgDependency)" "@(AnalyzerNupkgLibrary)" "$(_GeneratedRulesetsDir)" "$(_GeneratedEditorc ...
```
`csi.exe` is not available in the .NET 5 SDK that we are using for building. I ended up "porting" the `.csx` file to the `.cs` file. Where porting is just adding imports and wrapping the file contents in a `public static void Main(string[] Args)`.

omajid commented on patches/roslyn-analyzers/0001-WIP-for-source-build.patch:
```diff
@@ -0,0 +1,1087 @@
++  <PropertyGroup>
++    <OutputType>Exe</OutputType>
++    <TargetFramework>net5.0</TargetFramework>
++  </PropertyGroup>
```
Is there anything I need to do to make sure I am not accidentally using prebuilts here?

dagood replied:
In general, even the Microsoft build wants to be able to build a basic app without downloading everything (simple "offline" scenarios) so a clean/basic proj like this should be ok. Looking at the annotated prebuilt report for the `project.assets.json` of this proj is the way to tell from the end results.

omajid commented on patches/roslyn-analyzers/0001-WIP-for-source-build.patch:
```diff
@@ -0,0 +1,1087 @@
+       Restore would conclude that there is a cyclic dependency between Microsoft.CodeAnalysis and Microsoft.CodeAnalysis.Analyzers.
+     -->
+-    <PackageId>*$(MSBuildProjectFullPath)*</PackageId>
++    <PackageId>*$(MSBuildProjectFullPath.Replace('/', '\\'))*</PackageId>
```
As the comment in the original file says, they want to use the file path as the package name.

But when the "package name" (really the full file path with forward slashes on Linux) is put in the assets.json file, it ends up confusing the version parsing logic which thinks `"*/home/foo/bar/baz*/4.5.0"` has a version along the lines of `/home/foo/bar/baz*/4.5.0`.

Changing the slashes back to Windows-style slashes makes this work.

dagood replied:
Bizarre... can you add this to the code comment? (I imagine this can be upstreamed for compat, and would be nice to have explanation inline. If they don't want to find some other way.)

omajid replied:
Done.

omajid commented on patches/roslyn-analyzers/0001-WIP-for-source-build.patch:
```diff
@@ -0,0 +1,1087 @@
+\ No newline at end of file
++</Project>
+diff --git a/src/Microsoft.CodeAnalysis.Analyzers/Setup/AnalyzerReleases.Shipped.md b/src/Microsoft.CodeAnalysis.Analyzers/Setup/AnalyzerReleases.Shipped.md
+new file mode 100644
```
A standalone build of `roslyn-analyzers` was failing for me with messages like:

> /home/omajid/devel/dotnet/roslyn-analyzers/src/Directory.Build.targets(45,5): error : Create a new empty file named 'AnalyzerReleases.Shipped.md' in the project directory [/home/omajid/devel/dotnet/roslyn-analyzers/src/Roslyn.Diagnostics.Analyzers/Setup/Roslyn.Diagnostics.Analyzers.Setup.csproj]

So I did what it asked :shrug:

omajid commented on repos/roslyn-analyzers.proj:
```diff
@@ -0,0 +1,40 @@
+    <RepositoryReference Include="arcade" />
+    <RepositoryReference Include="command-line-api" />
+    <RepositoryReference Include="xliff-tasks" />
+    -->
```
I need to figure out the build dependencies of this. I am not sure how to do that.

dagood replied:
Basically, look at the prebuilt report and add any dependencies that would be resolved by other repos. (For non-nupkg dependencies you have to just know, but those are limited to a few repos.)

dagood commented on eng/Version.Details.xml:
```diff
@@ -50,6 +50,11 @@
       <Uri>https://github.com/dotnet/roslyn</Uri>
       <Sha>90e570d763637c4d0d30524674a64b07bb885994</Sha>
     </Dependency>
+    <!-- DOes this need a coherent parent dependency? -->
+    <Dependency Name="Microsoft.CodeAnalysis.NetAnalyzers" Version="5.0.0-rc1.20428.12">
```
Always, if possible--sdk has the dependency on roslyn-analyzers:

https://github.com/dotnet/sdk/blob/394fae2553937858c8f05a6158613dcc7f821fba/eng/Version.Details.xml#L176-L179

```xml
    <Dependency Name="Microsoft.CodeAnalysis.NetAnalyzers" Version="5.0.0-rc1.20428.12">
      <Uri>https://github.com/dotnet/roslyn-analyzers</Uri>
      <Sha>e856dc126950463c13226c7437fc7e9fb41442dc</Sha>
    </Dependency>
```

So here, we can do `CoherentParentDependency="Microsoft.NET.Sdk"` to link it up.

dagood commented on patches/roslyn-analyzers/0001-WIP-for-source-build.patch:
```diff
@@ -0,0 +1,1087 @@
++    {
++        public static void Main(string[] Args)
++        {
++            string nuspecFile = Args[0];
```
If you leave the original code unindented, would Git treat this as a file move + modify to be a bit more minimal? (And show if you had to make any tweaks?)

omajid replied:
Tried it out, doesn't seem to help. `git log -p -C` and `git show -C` don't show a diff.

The actual diff is: https://gist.github.com/omajid/e081771153fe7ea0fd4f23917a347e95

dagood replied:
Huh, I remember reading there's a heuristic with some percentage (90%?) of the file having to be the same to count as a copy/move. Oh well, this is fine. Thanks for the diff.

dagood commented on patches/roslyn-analyzers/0001-WIP-for-source-build.patch:
```diff
@@ -0,0 +1,1191 @@
+   </PropertyGroup>
+ 
++  <PropertyGroup>
++    <DotNetExecutible Condition="'$(OS)' == 'Windows_NT'">$(DotNetRoot)\dotnet.exe</DotNetExecutible>
++    <DotNetExecutible Condition="'$(DotNetExecutible)' == ''">$(DotNetRoot)/dotnet</DotNetExecutible>
```
(Nit) `DotNetExecutible` => `DotNetExecutable`

But come to think of it, this might be provided already by `$(DOTNET_HOST_PATH)`... can confirm in the binlog.

No need to reset CI just for this though.

dagood reviewed (approved):
LGTM when CI goes green--and heads up that it isn't running yet because there's a merge conflict (probably with the 5.0 GA PR).

omajid commented:
Thanks, but the CI wont be going green right now: this is dragging too many new prebuilts.

It looks like roslyn-analzyers depends on multiple versions of roslyn. For example, from a previous CI run:

```
/tb/tarball_874550/repos/Directory.Build.targets(553,5): error : Microsoft.CodeAnalysis.Common.1.2.1 [/tb/tarball_874550/repos/known-good.proj]
/tb/tarball_874550/repos/Directory.Build.targets(553,5): error : Microsoft.CodeAnalysis.Common.2.9.0 [/tb/tarball_874550/repos/known-good.proj]
/tb/tarball_874550/repos/Directory.Build.targets(553,5): error : Microsoft.CodeAnalysis.Common.3.3.1 [/tb/tarball_874550/repos/known-good.proj]
``` 

[Full log here](https://dev.azure.com/dnceng/public/_build/results?buildId=874550&view=logs&j=07b39231-98cb-5c72-d162-25ef5005a34b&t=82ce4082-533f-5b97-ca3f-d497b5f4c866&l=4083).

I will post a longer rant^Wdescription when I have rebased this to master.

omajid commented:
I am a bit stumped. Here's the new prebuilts after switching things from `netstandard1.3` to `netstandard2.0`:

```
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.3.3.1 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                                              
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.Analyzers.3.3.0 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                                    
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.BannedApiAnalyzers.3.3.0 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                           
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.Common.3.3.1 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                                       
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.CSharp.3.3.1 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                                       
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.CSharp.CodeStyle.3.8.0-2.20427.18 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                  
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.CSharp.Workspaces.3.3.1 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                            
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.FxCopAnalyzers.3.3.0 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                               
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.PublicApiAnalyzers.3.3.0 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                           
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.VersionCheckAnalyzer.3.3.0 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                         
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.VisualBasic.3.3.1 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                                  
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.VisualBasic.Workspaces.3.3.1 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                       
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.Workspaces.Common.3.3.1 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                            
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeQuality.Analyzers.3.3.0 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                                     
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.NetCore.Analyzers.3.3.0 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                                         
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.NetFramework.Analyzers.3.3.0 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]                                                                                    
/home/omajid/devel/dotnet/source-build/repos/Directory.Build.targets(553,5): warning : Roslyn.Diagnostics.Analyzers.3.3.0 [/home/omajid/devel/dotnet/source-build/repos/known-good.proj]
```

One issue is that this repo produces packages like `Roslyn.Diagnostics.Analzysers` version `3.3.1`, but the repo depends on [an older version of itself (`Roslyn.Diagnostics.Analyzers` version `3.3.0`)](https://github.com/dotnet/roslyn-analyzers/blob/03a1c1666000aad58d660db44806710f7e56326c/src/Directory.Build.props#L23). Any ideas how to break these dependencies? Any way to build `Roslyn.Diagnostics.Analyzers` from this repo and then use that to build everything else?

We also need packages (such as `Microsoft.CodeAnalysis.3.3.1` and `Microsoft.CodeAnalysis.Common`) that are produced by roslyn (I think).

dagood commented:
> One issue is that this repo produces packages like `Roslyn.Diagnostics.Analzysers` version `3.3.1`, but the repo depends on [an older version of itself (`Roslyn.Diagnostics.Analyzers` version `3.3.0`)](https://github.com/dotnet/roslyn-analyzers/blob/03a1c1666000aad58d660db44806710f7e56326c/src/Directory.Build.props#L23).

This seems reasonable--I saw this question on gitter before here and mentioned there:

> I suppose they're dogfooding... since they're analyzers I think we should be able to just get rid of them (since it should only be important when you're making changes to the code, not just compiling it)
>
> Otherwise we've got to bootstrap via previously-source-built, like Roslyn itself, I guess.
>
> So for your PR, I'd say don't worry about 3.3.0, we have to wait for an internal build with your changes so we can produce a previously-source-built tarball with 3.3.1, then we can force roslyn-analyzers to use 3.3.1 (to build 3.3.1 for downstream).

The dependency on dotnet/roslyn packages makes sense--it'd want to use the newest versions of the Roslyn APIs to poke around the code the analyze it.

About why they're showing up as prebuilts--dotnet/roslyn-analyzers might be missing the `eng/Version.props` -> `*.csproj` prop linkage that would let source-built override what they use with the source-built version of the package. A hard-coded version might be continuing to pull stuff down from online feeds.

omajid commented:
> A hard-coded version might be continuing to pull stuff down from online feeds.

There's some of that, for sure.

But it also appears like some rolsyn packages are not being built? `Microsoft.CodeAnalysis` 3.8.0 should have been build as part of roslyn repo, but it's not present. This appears to be a super-package; maybe I can patch roslyn-analzyers to use individual  nuget packages?

dagood commented:
> But it also appears like some rolsyn packages are not being built? `Microsoft.CodeAnalysis` 3.8.0 should have been build as part of roslyn repo, but it's not present. This appears to be a super-package; maybe I can patch roslyn-analzyers to use individual nuget packages?

Hmm. It: https://github.com/dotnet/roslyn/blob/4c195c3ac1974edcefa76774d7a59a2350ec55fa/src/NuGet/Microsoft.CodeAnalysis.Package.csproj
appears to be deliberately disabled: https://github.com/dotnet/roslyn/blob/4c195c3ac1974edcefa76774d7a59a2350ec55fa/src/NuGet/Directory.Build.props#L4

But I don't know why. 😕

It seems to me like it'd be reasonable to try patching `<ExcludeFromSourceBuild>false</ExcludeFromSourceBuild>` into that proj to get rid of the exclusion and try to build it.

@crummel @dseefeld any ideas?

omajid commented:
`Microsoft.CodeAnalysis` is also shipped with the SDK:

```
$ tar tf ~/downloads/dotnet-sdk-5.0.100-linux-x64.tar.gz | grep Microsoft.CodeAnalysis.dll
./sdk/5.0.100/Sdks/Microsoft.NET.Sdk.Razor/tools/Microsoft.CodeAnalysis.dll
./sdk/5.0.100/Roslyn/bincore/Microsoft.CodeAnalysis.dll
```

dagood commented:
Opened https://github.com/dotnet/source-build/issues/1869 to track down if that's an issue or not.

omajid commented:
The last commit brings the prebuilts down to:

```
source-build/repos/Directory.Build.targets(553,5): warning : 9 new packages used not in baseline! See report at source-build/artifacts/prebuilt-report/baseline-comparison.xml for more information. Package IDs are: [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.Analyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.BannedApiAnalyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.FxCopAnalyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.PublicApiAnalyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.VersionCheckAnalyzer.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeQuality.Analyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.NetCore.Analyzers.3.3.0 [source-build/repos/known-good.proj]                                                                                         
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.NetFramework.Analyzers.3.3.0 [source-build/repos/known-good.proj]                                                                                    
source-build/repos/Directory.Build.targets(553,5): warning : Roslyn.Diagnostics.Analyzers.3.3.0 [source-build/repos/known-good.proj]
```

The versions are the dogfood versions.

`Microsoft.CodeAnalysis.FxCopAnalyzers`, `Microsoft.CodeAnalysis.FxCopAnalyzers`, `Microsoft.NetCore.Analyzers`, and `Microsoft.NetFramework.Analyzers` are built by `FxCopAnalyzers.sln`. These are **not** part of the shipped SDK. I (accidentally) removed the `FxCopAnalyzsers.sln` early on; I am going to re-enable it now and see how things change.

dagood commented on patches/roslyn-analyzers/0003-Use-individual-Microsoft.CodeAnalysis-packages.patch:
```diff
@@ -0,0 +1,70 @@
++    <MicrosoftCodeAnalysisVisualBasicVersion>3.8.0</MicrosoftCodeAnalysisVisualBasicVersion>
+     <MicrosoftCodeAnalysisForRoslynDiagnosticsAnalyzersVersion>$(MicrosoftCodeAnalysisVersion)</MicrosoftCodeAnalysisForRoslynDiagnosticsAnalyzersVersion>
+-    <MicrosoftCodeAnalysisForShippedApisVersion>3.7.0</MicrosoftCodeAnalysisForShippedApisVersion>
++    <MicrosoftCodeAnalysisForShippedApisVersion>$(MicrosoftCodeAnalysisVersion)</MicrosoftCodeAnalysisForShippedApisVersion>
```
Heads up that this pattern *can* cause an issue if we need it to get updated by the PVP. Eval goes like:

1. `<MicrosoftCodeAnalysisVersion>3.8.0</>` -> set to 3.8.0
2. `<MicrosoftCodeAnalysisForShippedApisVersion>$(MicrosoftCodeAnalysisVersion)</>` -> set to 3.8.0 (fully evaluates now)
3. PVP overrides `MicrosoftCodeAnalysisVersion` based on `Microsoft.CodeAnalysis.*.nupkg`, say 3.8.1.
4. `MicrosoftCodeAnalysisForShippedApisVersion` is now "stranded" at 3.8.0

This can be fixed by adding one of these to make the PVP override both properties:

https://github.com/dotnet/source-build/blob/055775f3c4fd1120aeb69c404e76ce33a48cec05/Directory.Build.props#L255-L256

Depending on what this is used for, might be worth adding an `ExtraPackageVersionPropsPackageInfo` entry now. Shouldn't have any effect now, but might help us out with 5.0.101 or 102 if Roslyn gets updates.

omajid replied:
Thanks! Fixed now.

dagood commented on patches/roslyn-analyzers/0004-Giant-hacks.patch:
```diff
@@ -0,0 +1,92 @@
+     <!-- Use the correct compiler version -->
+     <UsingToolMicrosoftNetCompilers>true</UsingToolMicrosoftNetCompilers>
+     <!-- Roslyn -->
+-    <MicrosoftCodeAnalysisVersion>3.8.0</MicrosoftCodeAnalysisVersion>
++    <!-- Source build doesn't produce the Microsoft.CodeAnalysis super-assembly ; use individual packages instead
++    <MicrosoftCodeAnalysisVersion>3.3.1</MicrosoftCodeAnalysisVersion>
++    -->
```
"Metapackage" was the common term for this kind of package that has no content, only dependencies, when I was working with them in dotnet/runtime. And actually, since it doesn't have an assembly I'd expect it to be easy to add back. Filed https://github.com/dotnet/source-build/issues/1870 for this.

omajid commented:
`Microsoft.CodeAnalysis.FxCopAnalyzer` has a dependency (seen through `artifacts/obj/Microsoft.CodeAnalysis.FxCopAnalyzers.Setup/project.assets.json`) on `Microsoft.VisualStudio.Shell.*` packages, which are all proprietary.

It sounds like we need to break the build chain via SBRP at some point? How do we identify if it's at `Microsoft.CodeAnalysis.FxCopAnalyzers`+`Microsoft.CodeAnalysis.FxCopAnalyzers`+`Microsoft.NetCore.Analyzers` + `Microsoft.NetFramework.Analyzers` or at `Microsoft.VisualStudio.*`?

```
source-build/repos/Directory.Build.targets(553,5): warning : 35 new packages used not in baseline! See report at source-build/artifacts/prebuilt-report/baseline-comparison.xml for more information. Package IDs are: [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.Analyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.BannedApiAnalyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.FxCopAnalyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.PublicApiAnalyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeAnalysis.VersionCheckAnalyzer.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.CodeQuality.Analyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.NetCore.Analyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.NetFramework.Analyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.CoreUtility.15.6.27740 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.CoreUtility.16.0.428 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.ImageCatalog.16.0.28727 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Imaging.16.0.28729 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Imaging.Interop.14.0.DesignTime.14.3.26930 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.OLE.Interop.7.10.6071 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.SDK.EmbedInteropTypes.15.0.27 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Shell.15.0.16.0.28729 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Shell.Framework.16.0.28729 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Shell.Interop.7.10.6072 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Shell.Interop.10.0.10.0.30320 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Shell.Interop.11.0.11.0.61031 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Shell.Interop.12.0.12.0.30111 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Shell.Interop.14.0.DesignTime.14.3.26929 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Shell.Interop.15.3.DesignTime.15.0.26929 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Shell.Interop.15.6.DesignTime.15.6.27413 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Shell.Interop.8.0.8.0.50728 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Shell.Interop.9.0.9.0.30730 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Text.Data.16.0.428 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.TextManager.Interop.7.10.6071 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.TextManager.Interop.8.0.8.0.50728 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Threading.16.0.102 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Utilities.16.0.28729 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Microsoft.VisualStudio.Validation.15.3.15 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Roslyn.Diagnostics.Analyzers.3.3.0 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : StreamJsonRpc.1.5.43 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : System.Net.Http.4.3.3 [source-build/repos/known-good.proj]
source-build/repos/Directory.Build.targets(553,5): warning : Prebuilt usages are different from the baseline. If detected changes are acceptable, update baseline with: [source-build/repos/known-good.proj]
source-b
```

dagood commented:
Hmm, I would think that since `FxCopAnalyzers` doesn't end up in the SDK, we can stop building it, which is typically preferable. For anything else, there could be a problem where even though it's not in the SDK, it's executed during some repo build--but since this is an analyzer, we should be able to remove usages without affecting the build output.

(If we *did* need to execute this analyzer, we probably couldn't replace its dependencies with SBRP-built libraries anyway since they need to actually be runnable. I think in that case we'd be better off removing the dependency from the source and making whatever source removals we need so we can make sure it won't accidentally try to use an unavailable library.)

omajid commented:
Okay, I have reverted my changes to build FxCop and friends. And I have rebased this on top of https://github.com/dotnet/source-build/pull/1872.

CI says we are now back to:

```
/src/repos/Directory.Build.targets(553,5): warning : Prebuilt usages are different from the baseline. If detected changes are acceptable, update baseline with: [/src/repos/known-good.proj]
/src/repos/Directory.Build.targets(553,5): warning : cp '/src/artifacts/prebuilt-report/generated-new-baseline.xml' '/src/tools-local/prebuilt-baseline-online.xml' [/src/repos/known-good.proj]
/src/repos/Directory.Build.targets(553,5): error : 9 new packages used not in baseline! See report at /src/artifacts/prebuilt-report/baseline-comparison.xml for more information. Package IDs are: [/src/repos/known-good.proj]
/src/repos/Directory.Build.targets(553,5): error : Microsoft.CodeAnalysis.Analyzers.3.3.0 [/src/repos/known-good.proj]
/src/repos/Directory.Build.targets(553,5): error : Microsoft.CodeAnalysis.BannedApiAnalyzers.3.3.0 [/src/repos/known-good.proj]
/src/repos/Directory.Build.targets(553,5): error : Microsoft.CodeAnalysis.FxCopAnalyzers.3.3.0 [/src/repos/known-good.proj]
/src/repos/Directory.Build.targets(553,5): error : Microsoft.CodeAnalysis.PublicApiAnalyzers.3.3.0 [/src/repos/known-good.proj]
/src/repos/Directory.Build.targets(553,5): error : Microsoft.CodeAnalysis.VersionCheckAnalyzer.3.3.0 [/src/repos/known-good.proj]
/src/repos/Directory.Build.targets(553,5): error : Microsoft.CodeQuality.Analyzers.3.3.0 [/src/repos/known-good.proj]
/src/repos/Directory.Build.targets(553,5): error : Microsoft.NetCore.Analyzers.3.3.0 [/src/repos/known-good.proj]
/src/repos/Directory.Build.targets(553,5): error : Microsoft.NetFramework.Analyzers.3.3.0 [/src/repos/known-good.proj]
/src/repos/Directory.Build.targets(553,5): error : Roslyn.Diagnostics.Analyzers.3.3.0 [/src/repos/known-good.proj]
```

What are the next steps here?

Should I wait and rebase on top of https://github.com/dotnet/roslyn-analyzers/pull/4419 (and any followups)? Or is this PR okay to go for now?

dagood commented:
The steps are:
1. Grab `Logs centos71_Offline - Prebuilt Reports`.
1. Open it and copy the `generated-new-baseline.xml` content to `tools-local/...baseline...` (source-build = "online", tarball = "offline") to add entries that let this PR pass.
1. Merge PR (on us).
1. After the internal rolling build runs, we'll take the previously-source-built tar.gz it produces (prev-sb), upload it, and submit a PR to update the repo's prev-sb reference.
1. In that PR we can remove these baseline entries again since prev-sb provides them.

No need to wait for upstream, the requested changes look nonfunctional to me so far, and we'll have to resolve the patch conflicts later either way. Can fix them up later to make it a more obvious patch removal but we shouldn't wait on that--we need these prebuilts gone. 😄

omajid commented on patches/roslyn-analyzers/0001-WIP-for-source-build.patch:
```diff
@@ -0,0 +1,1467 @@
+From 48442474744a416eee4f367d3e389a0916047842 Mon Sep 17 00:00:00 2001
+From: Omair Majid <omajid@redhat.com>
+Date: Thu, 29 Oct 2020 18:20:09 -0400
+Subject: [PATCH 1/4] WIP for source-build
+
+---
```
Okay, but https://github.com/dotnet/roslyn-analyzers/pull/4419 is just a subset of this. That PR is missing `--pack` support, for example (I am working on a separate fix for that). There's also source-build specific fixes (which also should be upstreamed, but separately than 4419), and an `rm FxCopAnalyzers.sln` that's just not upstreamable.

Is there a good way I can clarify that some of this is under review and some of this is just ugly hacks to (somehow) get it to build?

omajid replied:
I edited the comment in this file to (hopefully) clarify this.

dagood replied:
I don't have any specific ideas here, I like what you did with it. Thanks. 🙂

omajid commented on tools-local/prebuilt-baseline-online.xml:
```diff
@@ -260,7 +269,7 @@
     <Usage Id="System.Diagnostics.FileVersionInfo" Version="4.3.0" />
     <Usage Id="System.Diagnostics.Process" Version="4.1.0" />
-    <Usage Id="System.Diagnostics.Process" Version="4.3.0" IsDirectDependency="true" />
+    <Usage Id="System.Diagnostics.Process" Version="4.3.0" />
```
Thanks! Should be fixed now.
